### PR TITLE
Add a few utility functions in Misc

### PR DESCRIPTION
**Pull request by mshinwell**
This GPR provides a few straightforward functions in `Misc` that are required by the forthcoming work on symbol types.  I don't think there is value in breaking this patch down into smaller pieces, given current time constraints.

### CODE REVIEW

File: utils/misc.ml [hunk 1]
```diff
@@ -214,6 +227,9 @@ module Stdlib = struct
         i = len || (f t.[i] && loop (i + 1))
       in
       loop 0
+
+    let print ppf t =
+      Format.pp_print_string ppf t
```
Comment (trefis) [inline utils/misc.ml:225]: Pffff.

File: utils/misc.ml [hunk 2]
```diff
@@ -163,6 +163,13 @@ module Stdlib = struct
       | None -> false
       | Some _ -> true
 
+    let compare f t1 t2 =
+      match t1, t2 with
+      | None, None -> 0
+      | None, Some _ -> -1
+      | Some _, None -> 1
+      | Some contents1, Some contents2 -> f contents1 contents2
```
Comment (trefis) [inline utils/misc.ml]: Would that function make sense in the stdlib?

**Review by trefis — COMMENTED** [review]
The `Option` module exists in the stdlib. Shouldn't you try using it, instead of `Misc.Stdlib.Option`?
Also, the one in `Misc` I believe duplicates some of the functions of the one in the stdlib. Perhaps it would be better to look into cleaning `Misc.Stdlib` instead of growing it even more.

**Comment by mshinwell** [comment]
Indeed, I think these probably predate the support in the `Stdlib` for `Option` that I had forgotten about.  I will see how much we can remove from `Misc`.

**Comment by mshinwell** [comment]
@trefis Please review again.  I have removed the functions from `Misc.Stdlib.Option` that are now in the stdlib.

**Review by trefis — CHANGES_REQUESTED** [review]
The cleanup of `Misc.Stdlib.Option` is nice!
Adding a printer in there seems reasonable.

The addition of `Misc.protect_writing_to_file` is also fine.

I would remove the alias to `Format.pp_print_string` from `Misc.Stdlib.String` though.

**Comment by trefis** [comment]
Also, I tend to agree with Alain: I'm not sure there is much value in a changelog entry for that PR.

**Comment by mshinwell** [comment]
The `print` function in `String` was put there for a reason: I'd like modules to have a uniform interface, which includes printing for all data structures.

**Comment by Drup** [comment]
> The `print` function in `String` was put there for a reason: I'd like modules to have a uniform interface, which includes printing for all data structures.

I agree with that on principle, the issue is dependency order of stdlib modules.

At some point, we need to have a serious though about the exact dependency orders we want for the stdlib modules. I remember @dbuenzli being pretty burned by that for the `UChar` module (iirc).

That being said, the convention for pretty printing functions is `pp`, not `print`. :)

**Comment by mshinwell** [comment]
@Drup Well, we need to have a serious discussion as to how printing can be added to the data structure modules in the stdlib.  Let's do that elsewhere though...

**Comment by dbuenzli** [comment]
> I remember @dbuenzli being pretty burned by that for the `UChar` module (iirc).

Just for the ref this was https://caml.inria.fr/mantis/view.php?id=7500

**Comment by trefis** [comment]
> I agree with that on principle, the issue is dependency order of stdlib modules.

We're not talking about the stdlib here, only of `Misc.Stdlib`.

Since people think it's fine I'll retract my comment and approve.

**Review by trefis — APPROVED** [review]
Feel free to merge once you've rebased.

**Pull request merged**